File: modules/core/src/main/java/com/panayotis/jubler/subs/loader/PlainText.java
/*
 * (c) 2005-2023 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

package com.panayotis.jubler.subs.loader;

import com.panayotis.jubler.subs.SubEntry;
import com.panayotis.jubler.time.Time;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PlainText extends AbstractTextSubFormat {

    private static final Pattern pat;
    private double current_time = 0;

    static {
        pat = Pattern.compile("(.*?)" + nl);
    }

    @Override
    public Pattern getPattern() {
        return pat;
    }

    @Override
    protected Pattern getTestPattern() {
        return getPattern();
    }

    protected SubEntry getSubEntry(Matcher m) {
        Time start = new Time(current_time);
        current_time += 2;
        Time finish = new Time(current_time);
        current_time += 1;
        return new SubEntry(start, finish, m.group(1));
    }

    public String getExtension() {
        return "txt";
    }

    public String getName() {
        return "PlainText";
    }

    @Override
    public String getExtendedName() {
        return "Plain text";
    }

    protected void appendSubEntry(SubEntry sub, StringBuilder str) {
        str.append(sub.getText()).append('\n');
    }

    @Override
    protected String initLoader(String input) {
        current_time = 0;
        return super.initLoader(input);
    }

    public boolean supportsFPS() {
        return false;
    }
}

File: src/core/formats/text/PlainText.cpp
/*
 * (c) 2005-2026 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

#include "core/formats/text/PlainText.h"

const QRegularExpression &PlainText::getPattern() {
    static const QRegularExpression pat(QStringLiteral("(.*?)") + nl);
    return pat;
}

SubEntryPtr PlainText::getSubEntry(const QRegularExpressionMatch &m) {
    const Time start(currentTime_);
    currentTime_ += 2;
    const Time finish(currentTime_);
    currentTime_ += 1;
    return std::make_shared<SubEntry>(start, finish, m.captured(1));
}

QString PlainText::initLoader(const QString &input) {
    currentTime_ = 0;
    return AbstractTextSubFormat::initLoader(input);
}

QList<SubEntryPtr> PlainText::loadSubtitles(const QString &input, bool debug) {
    QList<SubEntryPtr> entries = AbstractTextSubFormat::loadSubtitles(input, debug);
    // The decoder's artificial trailing line terminators must not become
    // empty subtitles.
    while (!entries.isEmpty() && entries.last()->getText().isEmpty())
        entries.removeLast();
    return entries;
}

void PlainText::appendSubEntry(const SubEntry &sub, QString &str) {
    str += sub.getText();
    str += QLatin1Char('\n');
}
